Title: Scheme/parsec-tpm: Add TPM vendor and model information

Sukuna0007Abhi opened the pull request:
Description:
This PR implements support for TPM vendor and model information in TPM endorsements, addressing Issue #139.

Changes include:
- Extended TaAttr structure with vendor and model fields
- Added input validation and sanitization for security
- Implemented comprehensive test coverage:
  * Real-world TPM vendor/model combinations
  * International character handling
  * Security test cases
  * Edge cases and input validation
- Updated documentation with new fields and security considerations

Security Considerations:
- Implemented strict input validation
- Added sanitization for potentially dangerous inputs
- Protected against injection attacks
- Enforced character and length limits

Testing:
- Added extensive test suite covering:
  * Common TPM vendors (Infineon, STMicroelectronics, etc.)
  * International vendor names (Unicode support)
  * Security test cases (injection attempts)
  * Edge cases and combinations
  * Character type handling

Documentation:
- Updated README.md with new field descriptions
- Added security considerations
- Included usage examples

Resolves: #139 

Ready for Review sir @thomas-fossati @setrofim @yogeshbdeshpande

Sukuna0007Abhi committed:
scheme/parsec-tpm: add TPM vendor and model information

This change implements support for TPM vendor and model information in TPM
endorsements (Issue #139). The implementation includes:

- Extended TaAttr structure with vendor and model fields
- Input validation and sanitization for security
- Comprehensive test coverage including:
  * Real-world TPM vendor/model combinations
  * International character handling
  * Security test cases
  * Edge cases and input validation
- Updated documentation with new fields and security considerations

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

setrofim commented on scheme/parsec-tpm/README.md:
```diff
@@ -24,7 +24,116 @@
+      "key": [{
+        "type": "pkix-base64-key",
+        "value": "MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAETKRFE_RwSXooI8DdatPOYg_uiKm2XrtT_uEMEvqQZrwJHHcfw0c3WVzGoqL3Y_Q6xkHFfdUVqS2WWkPdKO03uw==",
+        "parameters": {
```
`crypto-key-type-choice` offers "type choice" rather than "map" extension point. Meaning you cannot add fields to existing values, you can only add new types. So you cannot add parameters to, e.g., `"pkix-base64-key"`, you'd need to define a whole new type, `"pkix-base64-key-with-params"` and register that as an extension to the type choice (and then do the same for every other variant that needs to be extended in this way).

Sukuna0007Abhi replied:
Ready for review sir

Sukuna0007Abhi committed:
Refactor: Move vendor/model to environment.class per CoRIM spec

Following @setrofim's feedback that crypto-key-type-choice uses
'type choice' not 'map' extension point, refactored to comply with
standard CoRIM specification.

Changes:
- REMOVED: key.Parameters approach (invalid per CoRIM spec)
- ADDED: Extract vendor/model from environment.class (standard location)
- Follows same pattern as PSA scheme
- Fixed regex backreference issues in sanitization
- Updated README with complete, valid CoRIM examples
- Deleted ta_attrs.go and related test files (wrong approach)
- All tests passing

This addresses the review comment that we cannot add parameters
to existing key types like 'pkix-base64-key'. Instead, vendor/model
are now stored in the standard environment.class location.

Resolves: CoRIM specification compliance issue

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

setrofim reviewed (approved):
LGTM

yogeshbdeshpande commented on scheme/parsec-tpm/sanitize.go:
```diff
@@ -0,0 +1,213 @@
+// Copyright 2024 Contributors to the Veraison project.
```
```suggestion
// Copyright 2025 Contributors to the Veraison project.
```

yogeshbdeshpande commented on scheme/parsec-tpm/sanitize_test.go:
```diff
@@ -0,0 +1,125 @@
+// Copyright 2024 Contributors to the Veraison project.
```
```suggestion
// Copyright 2025 Contributors to the Veraison project.
```

yogeshbdeshpande commented on scheme/parsec-tpm/sanitize.go:
```diff
@@ -0,0 +1,213 @@
+// Copyright 2024 Contributors to the Veraison project.
```
I am wondering this generic implementation of string validation and checking for specific patterns is applicable more generically across entire Services `scheme` and could be moved to `services/common` ??

yogeshbdeshpande replied:
ok, understood, thanks. Are these then coming from a spec reference for restriction in TPM specific scene setting?

setrofim replied:
Not that I'm aware of (as I've said -- I'm not familiar with the specs). But this looks like it's designed specifically to catch "unreasonable" vendor/model names. From that point of view, the implementation looks reasonable. But I would not accept this as a good generic santizer (not do I think it's necesarily a good idea to have one).

yogeshbdeshpande reviewed (commented):
Please check my comments!

Sukuna0007Abhi committed:
Update scheme/parsec-tpm/sanitize.go

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Co-authored-by: Yogesh Deshpande <yogesh.deshpande@arm.com>
Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi committed:
Update scheme/parsec-tpm/sanitize_test.go

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Co-authored-by: Yogesh Deshpande <yogesh.deshpande@arm.com>
Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi commented:
Sure sir, @yogeshbdeshpande will try my best, to explain properly 🙏

yogeshbdeshpande commented on scheme/parsec-tpm/README.md:
```diff
@@ -24,7 +24,166 @@
+- `parsec-tpm.vendor`: The manufacturer or vendor of the TPM (optional)
+- `parsec-tpm.model`: The specific model identifier of the TPM (optional)
+
+Both fields support:
+- Unicode characters (e.g., for international vendor names)
+- Special characters (allowed in both fields)
+- Variable length strings (no artificial length restrictions)
```
```suggestion
Both fields support:
- Unicode characters (e.g., for international vendor names)
- Special characters (allowed in both fields)
- Variable length strings (no artificial length restrictions)
```

yogeshbdeshpande commented on scheme/parsec-tpm/README.md:
````diff
@@ -24,7 +24,166 @@
+  ]
+}
+```
+
````
I do not think we need additional example, just drop them

yogeshbdeshpande commented on scheme/parsec-tpm/corim_extractor.go:
```diff
@@ -121,6 +122,27 @@ func makeTaAttrs(id ID, key *comid.CryptoKey) (json.RawMessage, error) {
+		if env.Class.Vendor != nil {
+			vendor := string(*env.Class.Vendor)
+			// Trim and validate
+			vendor = sanitizeAndValidate(vendor)
```
Line 131 not needed!

yogeshbdeshpande commented on scheme/parsec-tpm/corim_extractor.go:
```diff
@@ -121,6 +122,27 @@ func makeTaAttrs(id ID, key *comid.CryptoKey) (json.RawMessage, error) {
+		if env.Class.Model != nil {
+			model := string(*env.Class.Model)
+			// Trim and validate
+			model = sanitizeAndValidate(model)
```
Line 139 not needed!

yogeshbdeshpande commented on scheme/parsec-tpm/corim_extractor.go:
```diff
@@ -188,3 +210,22 @@ func (o *ID) FromEnvironment(e comid.Environment) error {
 func (o *CorimExtractor) SetProfile(profile string) {
 	o.Profile = profile
 }
+
```
Remove Lines 213 to 231

yogeshbdeshpande commented on scheme/parsec-tpm/sanitize.go:
```diff
@@ -0,0 +1,213 @@
+// Copyright 2025 Contributors to the Veraison project.
```
File can be deleted!

yogeshbdeshpande reviewed (changes requested):
See my all comments, please re-work on the PR!

Sukuna0007Abhi committed:
Simplify vendor/model implementation per review feedback

- Remove sanitization files (sanitize.go, sanitize_test.go)
- Extract vendor/model directly without validation
- Simplify README: keep one example, remove Security section
- Add vendor/model support to SwAttr for supply chain

Addresses feedback from sir @yogeshbdeshpande

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>
